File: solutions/3500-3599/3515.shortest-path-in-a-weighted-tree/Solution.java
import java.util.*;

class FenwickTree {
    private int size;
    private long[] tree;

    public FenwickTree(int size) {
        this.size = size;
        this.tree = new long[size + 1];
    }

    public void update(int index, long delta) {
        for (; index <= size; index += index & -index) {
            tree[index] += delta;
        }
    }

    public long query(int index) {
        long sum = 0;
        for (; index > 0; index -= index & -index) {
            sum += tree[index];
        }
        return sum;
    }
}

class Tree {
    private int n;
    private List<List<int[]>> adj;
    private int[] tin, tout, parent, depth;
    private long[] rootDistance;
    private int timer;

    public Tree(int n) {
        this.n = n;
        this.adj = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }
        this.tin = new int[n + 1];
        this.tout = new int[n + 1];
        this.parent = new int[n + 1];
        this.depth = new int[n + 1];
        this.rootDistance = new long[n + 1];
    }

    public void addEdge(int u, int v, int w) {
        adj.get(u).add(new int[] { v, w });
        adj.get(v).add(new int[] { u, w });
    }

    public void initialize(int root) {
        this.timer = 0;
        dfs(root, 0, 0L);
    }

    private void dfs(int u, int p, long dist) {
        parent[u] = p;
        rootDistance[u] = dist;
        tin[u] = ++timer;
        depth[u] = depth[p] + 1;

        for (int[] edge : adj.get(u)) {
            int v = edge[0], w = edge[1];
            if (v != p) {
                dfs(v, u, dist + w);
            }
        }

        tout[u] = timer;
    }

    public int getInTime(int u) {
        return tin[u];
    }

    public int getOutTime(int u) {
        return tout[u];
    }

    public long getRootDistance(int u) {
        return rootDistance[u];
    }

    public int getParent(int u) {
        return parent[u];
    }

    // New method to get edges for a node
    public List<int[]> getEdges(int u) {
        return adj.get(u);
    }
}

class Solution {
    public int[] treeQueries(int n, int[][] edges, int[][] queries) {
        Tree tree = new Tree(n);

        // Build the tree
        for (int[] edge : edges) {
            tree.addEdge(edge[0], edge[1], edge[2]);
        }
        tree.initialize(1);

        FenwickTree fenwick = new FenwickTree(tree.getOutTime(1));
        int[] upEdgeWeight = new int[n + 1];

        // Initialize upEdgeWeight array using the new getEdges method
        for (int u = 2; u <= n; u++) {
            for (int[] edge : tree.getEdges(u)) {
                if (edge[0] == tree.getParent(u)) {
                    upEdgeWeight[u] = edge[1];
                    break;
                }
            }
        }

        List<Integer> result = new ArrayList<>();

        for (int[] query : queries) {
            if (query[0] == 1) {
                // Update operation
                int u = query[1], v = query[2], newWeight = query[3];
                int child = (tree.getParent(u) == v) ? u : v;

                long delta = newWeight - upEdgeWeight[child];
                fenwick.update(tree.getInTime(child), delta);
                fenwick.update(tree.getOutTime(child) + 1, -delta);
                upEdgeWeight[child] = newWeight;
            } else {
                // Query operation
                int u = query[1];
                long currentDistance = tree.getRootDistance(u) + fenwick.query(tree.getInTime(u));
                result.add((int) currentDistance);
            }
        }

        return result.stream().mapToInt(i -> i).toArray();
    }
}

File: solutions/3500-3599/3515.shortest-path-in-a-weighted-tree/Solution.cpp
struct Fenwick
{
    int n;
    vector<long long> BIT;
    Fenwick(int n) : n(n), BIT(n + 1, 0) {}
    void update(int i, long long delta)
    {
        for (; i <= n; i += i & -i)
            BIT[i] += delta;
    }
    long long query(int i)
    {
        long long sum = 0;
        for (; i; i -= i & -i)
            sum += BIT[i];
        return sum;
    }
};

struct Tree
{
    int N, root = 1;
    vector<vector<pair<int, int>>> adj;
    vector<int> tin, tout, par, depth;
    vector<long long> rootD;
    int timer = 0;
    Tree(int _N)
    {
        N = _N;
        adj.resize(N + 1);
        tin.resize(N + 1);
        tout.resize(N + 1);
        par.resize(N + 1);
        depth.resize(N + 1);
        rootD.resize(N + 1);
    }
    void add_edge(int u, int v, int w)
    {
        adj[u].push_back({v, w});
        adj[v].push_back({u, w});
    }
    void init(int rt = 1)
    {
        root = rt;
        timer = 0;
        function<void(int, int, long long)> dfs = [&](int u, int p, long long dist)
        {
            par[u] = p;
            rootD[u] = dist;
            tin[u] = ++timer;
            depth[u] = depth[p] + 1;
            for (auto &[v, w] : adj[u])
                if (v != p)
                    dfs(v, u, dist + w);
            tout[u] = timer;
        };
        dfs(root, 0, 0);
        // DFS Preprocessing for calcualting parent , rootD , and tin and tout times
    }
};

class Solution
{
public:
    vector<int> treeQueries(int n, vector<vector<int>> &edges, vector<vector<int>> &queries)
    {
        Tree tree(n);
        vector<int> ans;
        for (auto &e : edges)
            tree.add_edge(e[0], e[1], e[2]);
        tree.init(1);
        Fenwick F(tree.timer);
        vector<int> up(n + 1, 0);
        for (int i = 2; i <= n; ++i)
            for (auto &[u, w] : tree.adj[i])
                if (u == tree.par[i])
                    up[i] = w;
        for (auto &q : queries)
        {
            if (q[0] == 1)
            {
                int u = q[1], v = q[2], c = v;
                if (tree.par[u] == v)
                    c = u; // c = child
                // Range Update in Fenwick Tree , the change in weight
                // delta = q[3] - up[c]
                // we add this value at tin[child]
                // and subtract at tout[child] + 1
                F.update(tree.tin[c], q[3] - up[c]);
                F.update(tree.tout[c] + 1, up[c] - q[3]);
                up[c] = q[3];
            }
            else
                ans.push_back((int)(tree.rootD[q[1]] + F.query(tree.tin[q[1]])));
            //                             intial weight   + changes done in weights
        }
        return ans;
    }
};
